### PR TITLE
Fixed bug in sf::Text when applying an outline color/thickness

### PR DESCRIPTION
**Pull request by assematt**
Link to the [thread](http://en.sfml-dev.org/forums/index.php?topic=21246.0) on the forum.

When applying an outline thickness to sf::Text in combination with a
strikethrough and/or an underlined style, the ensureGeometryUpdate
function adds unwanted vertices if the string contains two consecutive
'\n' charecter.
To fix this we need to add an additional check in the if statements to
check if both the current and previous character it's a new line
character.

This is a visual representation of the bug
textbug

And this is the code that generated it.

`sf::Text TestString;`
`TestString.setFont(TestFont);`
`TestString.setCharacterSize(60u);`
`TestString.setOutlineColor(sf::Color::Red);`
`TestString.setOutlineThickness(2.f);`
`TestString.setStyle(sf::Text::StrikeThrough | sf::Text::Underlined);`
`TestString.setString("Line 1\n\nLine 2");`

### CODE REVIEW

**Comment by eXpl0it3r** [comment]
Thanks for the pull request! Keep in mind that we usually like to discuss issues and features on the [SFML forum](http://en.sfml-dev.org/forums/index.php) as stated in the [contribution guidelines](http://www.sfml-dev.org/contribute.php).

The if blocks now only get entered if there are two consecutive new line characters and if there's just one it won't get entered. Is that really the correct behavior?

In general (not directly related to this PR) I just wondered does SFML only support \n as new line character? What about \r and \l?

**Comment by assematt** [comment]
My bad for the lack of posting this before in the SFML forum, so here the link to the [thread](http://en.sfml-dev.org/forums/index.php?topic=21246.0) on the forum.

Also you are right, I made a mistake when copying (or in this case lack of ctrl-c/ctrl-v) the fixed code from my custom text class to the standard one. The correct check is ... `&& (prevChar != L'\n')`. We should enter the if statement if the current char it's a new line charcter and the previous wasn't. My bad again.

**Comment by assematt** [comment]
I updated the code to better support \r\n as new line char without a graphic glitch.
Now sf::Text should not produce unwanted graphics glitches when using an outline color/thickness if the sf::Text string contains two consecutive new line character. A posted in the [thread](http://en.sfml-dev.org/forums/index.php?topic=21246.msg151667#msg151667) on the forum some comparison image to better understand the current problem and the fix.

**Comment by MarioLiebisch** [comment]
As mentioned on the forums, I think `\r` should be treated as a real carriage return:

> I actually think "\r" shouldn't be ignored and instead be processed as expected (i.e. reset x position for next glyph).
>
> The common use case in Windows line breaks won't be affected, but it would also make it possible to actually utilize it the way it's meant to be (return to the start of the line without beginning a new line).

I'm aware that this might be unexpected for Mac users, but then again there's no real way to deal with this effectively.

**Comment by eXpl0it3r** [comment]
@SFML/sfml Should `\r` be treated as dedicated line break? What about in combination `\r\n`?

**Comment by binary1248** [comment]
`\r` isn't a line break... It's a carriage return. For some stupid reason Apple decided to re-purpose it as a line-feed well aware that there was already a character for that, `\n`. This is one of those murky places that nobody really likes to visit.

Since SFML is meant to help people write cross-platform code, we have to define some form of standard that will produce the same behaviour everywhere. We really shouldn't care what the "native" line break looks like on each platform, because quite frankly, the developer using SFML wouldn't want to care themselves anyway. If they want a line break (with implicit carriage return) in their `sf::Text` they should just have to insert a `\n` somewhere. We can also support `\r` as a carriage return without line feed, but I wouldn't try to combine it into CRLF like it is done on Windows, nor would I interpret it as a full line break like it used to be done by Apple. The Unix way is the one true way, and that is also what is used by the C/C++ standards when using line break translation (i.e. talking to file streams opened in text mode).

**Comment by binary1248** [comment]
I think handling `'\r'` in the context of `sf::Text` makes little to no sense. We don't have a typewriter head and the user is free to position the drawable at any pixel position they desire. I really cannot imagine anybody wanting to overdraw previous characters on the same line by (mis)using `'\r'`.

@assematt Can you remove the `'\r'` handling, squash and push a new version of the change?

**Comment by eXpl0it3r** [comment]
And replace tabs with spaces for indenting and rebase to resolve the merge conflict. 😄

**Comment by eXpl0it3r** [comment]
Ping @assematt 🙂

**Comment by assematt** [comment]
Hello everyone. I updated the code with the updated handling of the '\r' character. Everything should work now without any weird glitches. The new code will produce a new line only when the '\n' character is encountered and will ignore any '\r' character.

**Comment by LaurentGomila** [comment]
I see tabs in your latest commit, they should be replaced with spaces.

**Comment by assematt** [comment]
Yep, my bad. Somehow some tabs escaped the search&replace in notepad++ :D Should be fixed now

File: src/SFML/Graphics/Text.cpp [hunk 1]
```diff
@@ -443,13 +443,16 @@ void Text::ensureGeometryUpdate() const
     for (std::size_t i = 0; i < m_string.getSize(); ++i)
     {
         Uint32 curChar = m_string[i];
+        
+        // Skip the /r char to avoid weird graphical issues
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/Text.cpp]: Should also be \r in the comment

**Comment by eXpl0it3r** [comment]
Squashed and merged in bcb013b45b89528608809bdd68b6a86cd0949b8c

Thanks for the contribution! 🙂

**Pull request closed without merging**